File: src/PAhAom/GameState/Menu.cpp
#include <PAhAom/GameState/GameState.hpp>
#include <PAhAom/Resources.hpp>
#include <PAhAom/Settings.hpp>

#include <util/Observer.hpp>


namespace PAhAom::GameState
{

Menu::Menu()
: text("Menu, press Enter to continue,\nEsc to exit.", *Resources::get<sf::Font>("unifont"))
{
        this->text.setPosition(sf::Vector2f(Settings::Video::resolution) / 2.f);
        this->text.setOrigin({ this->text.getGlobalBounds().width  / 2.f,
                               this->text.getGlobalBounds().height / 2.f });
}

void Menu::handleInput(const sf::Event& event)
{
        if (event.type == sf::Event::KeyPressed)
        {
                switch (event.key.code)
                {
                        case sf::Keyboard::Return:
                        {
                                util::Subject::send({ util::Message::PushState{ new Game() }});
                                break;
                        }

                        case sf::Keyboard::Escape:
                        {
                                util::Subject::send({ util::Message::PopState{} });
                                break;
                        }

                        default: break;
                }
        }
}

void Menu::update(const util::DeltaTime dt)
{

}

void Menu::draw(sf::RenderTarget& target)
{
        target.draw(this->text);
}

}

File: data/scripts/GameState/Menu.lua
Menu = {
    entities = {
        {
            position = {
                x = Settings.video.resolution.x / 2,
                y = Settings.video.resolution.y / 2,
            },
            graphics2D = {
                type = "text",
                content = "Menu, press Enter to continue,\nEsc to exit.",
                font = "unifont",
                fillColor = "yellow",
                origin = "center"
            },
        },
    }
}

function Menu:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end

function Menu:handle_input(event)
    if event.type == "KeyPressed" then
        local kc = event.code
        if kc == Keyboard.Enter then
            push_state "Game"
        elseif kc == Keyboard.Escape then
            pop_state()
        end
    end
end

function Menu:update(dt)
end

function Menu:draw()
    return self.entities
end